File: task-1/frontend/src/services/userService.ts
import api from './api';
import {
  UserData,
  UserCreateData,
  UserUpdateData,
  PasswordResetData,
  UserListResponse,
} from '../types/user';

export const userService = {
  async listUsers(
    page = 1,
    limit = 10,
    search?: string
  ): Promise<UserListResponse> {
    const params = new URLSearchParams({
      page: page.toString(),
      limit: limit.toString(),
    });
    if (search) params.append('search', search);
    const response = await api.get<UserListResponse>(
      `/users?${params.toString()}`
    );
    return response.data;
  },

  async getUser(id: number): Promise<UserData> {
    const response = await api.get<UserData>(`/users/${id}`);
    return response.data;
  },

  async createUser(
    user: UserCreateData,
    sendEmail = false
  ): Promise<UserData> {
    const response = await api.post<UserData>(
      `/users?send_email=${sendEmail}`,
      user
    );
    return response.data;
  },

  async updateUser(id: number, user: UserUpdateData): Promise<UserData> {
    const response = await api.put<UserData>(`/users/${id}`, user);
    return response.data;
  },

  async deleteUser(id: number): Promise<void> {
    await api.delete(`/users/${id}`);
  },

  async resetPassword(
    id: number,
    data: PasswordResetData,
    sendEmail = false
  ): Promise<{ message: string; email_sent: boolean }> {
    const response = await api.post(
      `/users/${id}/reset-password?send_email=${sendEmail}`,
      data
    );
    return response.data;
  },

  async sendPassword(
    id: number,
    data: PasswordResetData
  ): Promise<{ message: string; email_sent: boolean }> {
    const response = await api.post(`/users/${id}/send-password`, data);
    return response.data;
  },
};

File: task-1/backend/app/services/user_service.py
import logging
from sqlalchemy.orm import Session
from sqlalchemy import or_
from math import ceil
from typing import Optional
from ..models.user import User
from ..utils.security import get_password_hash

logger = logging.getLogger(__name__)


def create_user(
    db: Session, username: str, email: str, password: str, full_name: str, role: str
) -> User:
    """Create a new user"""
    password_hash = get_password_hash(password)
    user = User(
        username=username,
        email=email,
        password_hash=password_hash,
        full_name=full_name,
        role=role,
    )
    db.add(user)
    db.commit()
    db.refresh(user)
    return user


def update_user(db: Session, user_id: int, **kwargs) -> Optional[User]:
    """Update user fields"""
    user = db.query(User).filter(User.id == user_id).first()
    if not user:
        return None
    for key, value in kwargs.items():
        if value is not None:
            setattr(user, key, value)
    db.commit()
    db.refresh(user)
    return user


def reset_password(db: Session, user_id: int, new_password: str) -> Optional[User]:
    """Reset user password"""
    user = db.query(User).filter(User.id == user_id).first()
    if not user:
        return None
    user.password_hash = get_password_hash(new_password)
    db.commit()
    db.refresh(user)
    return user


def delete_user(db: Session, user_id: int) -> Optional[User]:
    """Soft delete a user"""
    user = db.query(User).filter(User.id == user_id).first()
    if not user:
        return None
    user.is_active = False
    db.commit()
    return user


def list_users(
    db: Session, page: int = 1, limit: int = 10, search: Optional[str] = None
):
    """List users with pagination and search"""
    query = db.query(User)

    if search:
        search_term = search.strip()
        if search_term:
            search_filter = or_(
                User.username.ilike(f"%{search_term}%"),
                User.full_name.ilike(f"%{search_term}%"),
                User.email.ilike(f"%{search_term}%"),
            )
            query = query.filter(search_filter)

    total = query.count()
    offset = (page - 1) * limit
    users = query.order_by(User.created_at.desc()).offset(offset).limit(limit).all()
    total_pages = ceil(total / limit) if limit > 0 else 0

    return {
        "total": total,
        "page": page,
        "limit": limit,
        "total_pages": total_pages,
        "data": users,
    }
